File: vivial_connect_example.js
/* 
 Requires node version 6+
 Install dependencies (request, moment, crypto-js, url)
 Fill in the empty constants (ln 15 - ln 19)
 Execute the script with the following command: node vivial_connect_example.js
 This will send a message to the 'to_number' you provided
*/

'use strict';

const request = require('request');
const moment = require('moment');
const cryptoJS = require('crypto-js');
const url_lib = require('url').URL;

const base_url = 'https://api.vivialconnect.net/api/v1.0';
const canonicalized_header_names = 'accept;date;host';
const api_key = '';
const api_secret = '';
const account_id = '';
const to_number = '';
const from_number = '';

function signRequest(method, url, data){

    var now = moment.utc();
    var requestTimestamp = now.format('YYYYMMDD[T]HHmmss[Z]');
    var dateForHeader = now.format('ddd, DD MMM YYYY HH:mm:ss [GMT]');
    var SHA256data = cryptoJS.SHA256(data);
    var myURL = new url_lib(url);
    var path = myURL.pathname;
    var sortedParams = '';
    var canonicalizedHeaders = 'accept:application/json' + '\n' + `date:${dateForHeader}` + '\n' +'host:api.vivialconnect.net';

    var canonicalRequest = method + '\n' + 
                         requestTimestamp + '\n' +
                         path + '\n' +
                         sortedParams + '\n' +
                         canonicalizedHeaders + '\n' +
                         canonicalized_header_names + '\n' +
                         SHA256data;


    var signature = cryptoJS.HmacSHA256(canonicalRequest, api_secret);

    var signatureData = {
        'dateForHeader': dateForHeader,
        'requestTimestamp': requestTimestamp,
        'signature': signature
    };

    return signatureData
}

function makeRequest(method, url, data) {
    var signatureData = signRequest(method, url, data);
    var options = {
        'url': url,
        'method': method,
        'headers': {
            'Accept': 'application/json',
            'Date': signatureData.dateForHeader,
            'Host': 'api.vivialconnect.net',
            'Content-Type': 'application/json',
            'X-Auth-Date': signatureData.requestTimestamp,
            'X-Auth-SignedHeaders': 'accept;date;host',
            'Authorization': 'HMAC ' + api_key + ':' + signatureData.signature
        },
        body: data
    };

    request(options, function (error, response, body) {
        console.log('error:', error)
        console.log('statusCode:', response && response.statusCode)
        console.log('body:', body)
    });
}

(function(){
    var data = { 
        'message': {
            'to_number': to_number, 
            'from_number': from_number, 
            'body': 'Hello from the Vivial Connect Node.js example'
        }
    };

    data = JSON.stringify(data)
    var url = `${base_url}/accounts/${account_id}/messages.json`
    makeRequest('POST', url , data)
})();

File: vivial_connect_example.rb
# Install required gems
# Fill in empty constants (ln 14 - ln 18)
# Execute the script with the following command: ruby vivial_connect_example.rb
# This will send a message to the 'to_number' you provided

require 'json'
require 'faraday'
require 'digest'
require 'addressable'
require 'openssl'
require 'pp'

BASE_API_PATH = 'https://api.vivialconnect.net/api/v1.0/'
API_KEY =
API_SECRET =
ACCOUNT_ID =
TO_NUMBER =
FROM_NUMBER =


def build_canonical_request(http_verb, url, request_timestamp, canonicalized_headers, canonicalized_header_names, data={})
  canonicalized_query_string = ''
  canonical_request = http_verb + "\n" + request_timestamp + "\n" + Addressable::URI.encode(url.path) +
                      "\n" + canonicalized_query_string + "\n" + canonicalized_headers + "\n" +
                       canonicalized_header_names +  "\n" + Digest::SHA256.hexdigest(data.to_s)
end

def sign_request(http_verb, url, data={})
  set_request_timestamp
  set_date_for_date_header
  canonicalized_headers = "accept:application/json" + "\n" + "date:#{date_for_date_header}" + "\n" +"host:api.vivialconnect.net"
  canonicalized_header_names = "accept;date;host"
  canonical_request = build_canonical_request(http_verb, url, request_timestamp, canonicalized_headers, canonicalized_header_names, data)
  set_hmac_sha256(canonical_request)
end

def create_request_headers
  headers = {}
  headers['Content-Type'] = 'application/json'
  headers['Host'] = 'api.vivialconnect.net'
  headers['X-Auth-Date'] = request_timestamp
  headers['X-Auth-SignedHeaders'] = 'accept;date;host'
  headers['Authorization'] = "HMAC" + " " + API_KEY + ":" + hmac_sha256
  headers['Date'] = date_for_date_header
  headers['Accept'] = 'application/json'
  headers
end

def make_request(request_method, url, data={})
  base_path = BASE_API_PATH + "accounts/#{ACCOUNT_ID}"
  url = Addressable::URI.parse(base_path + url)
  sign_request(request_method, url, data)
  headers = create_request_headers
  request_method = request_method.downcase.to_sym
  connection = Faraday.new
  response = connection.run_request(request_method, url, data.to_s, headers)
end

def set_hmac_sha256(canonical_request)
  @hmac_sha256 = OpenSSL::HMAC.hexdigest('SHA256', API_SECRET, canonical_request)
end

def hmac_sha256
  @hmac_sha256
end

def set_request_timestamp
  @request_timestamp = Time.now.utc.strftime('%Y%m%dT%H%M%SZ')
end

def request_timestamp
  @request_timestamp
end

def set_date_for_date_header
  @date_for_date_header = Time.now.utc.strftime('%a, %d %b %Y %H:%M:%S GMT')
end

def date_for_date_header
  @date_for_date_header
end

def message_data(to_number, from_number)
  data = {}
  data['message'] = {}
  data['message']['to_number'] = to_number
  data['message']['from_number'] = from_number
  data['message']['body'] = "Hello World, from Vivial Connect"
  data.to_json
end

data = message_data(TO_NUMBER, FROM_NUMBER)
send_message_path = "/messages.json"
response = make_request("POST", send_message_path, data)
puts response.marshal_dump
